File: practical-10-row-transposition-cipher/row_transposition_cipher.py
def get_key_order(keyword):
    """Get column order based on alphabetical sorting of keyword."""
    keyword = keyword.upper().replace(" ", "")

    indexed = [(char, i) for i, char in enumerate(keyword)]
    sorted_indexed = sorted(indexed, key=lambda x: (x[0], x[1]))

    order = [0] * len(keyword)
    for new_pos, (_, old_pos) in enumerate(sorted_indexed):
        order[old_pos] = new_pos

    return order


def row_transposition_encrypt(plaintext, keyword):
    """Encrypt plaintext using Row Transposition cipher."""
    keyword = keyword.upper().replace(" ", "")
    key_len = len(keyword)

    padding_needed = (key_len - len(plaintext) % key_len) % key_len
    plaintext = plaintext + "X" * padding_needed

    rows = []
    for i in range(0, len(plaintext), key_len):
        rows.append(list(plaintext[i : i + key_len]))

    order = get_key_order(keyword)
    encrypted = ""

    for col_idx in range(key_len):
        original_col = order.index(col_idx)
        for row in rows:
            if col_idx < len(row):
                encrypted += row[original_col]

    return encrypted


def row_transposition_decrypt(ciphertext, keyword):
    """Decrypt ciphertext from Row Transposition cipher."""
    keyword = keyword.upper().replace(" ", "")
    key_len = len(keyword)
    num_rows = (len(ciphertext) + key_len - 1) // key_len

    order = get_key_order(keyword)

    grid = [[""] * key_len for _ in range(num_rows)]

    char_idx = 0
    sorted_cols = sorted(range(key_len), key=lambda x: (keyword[x], x))

    for new_col, original_col in enumerate(sorted_cols):
        for row in range(num_rows):
            if char_idx < len(ciphertext):
                grid[row][original_col] = ciphertext[char_idx]
                char_idx += 1

    plaintext = ""
    for row in grid:
        plaintext += "".join(row)

    return plaintext.rstrip("X")


if __name__ == "__main__":
    text = "ATTACK AT DAWN"
    key = "ZEBRAS"

    encrypted = row_transposition_encrypt(text, key)
    decrypted = row_transposition_decrypt(encrypted, key)

    print(f"Original: {text}")
    print(f"Key: {key}")
    print(f"Encrypted: {encrypted}")
    print(f"Decrypted: {decrypted}")

File: c programs/row_transposition_cipher.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

void get_key_order(const char *keyword, int order[]) {
    int key_len = strlen(keyword);
    
    for (int i = 0; i < key_len; i++) {
        order[i] = i;
    }
    
    for (int i = 0; i < key_len - 1; i++) {
        for (int j = i + 1; j < key_len; j++) {
            char ci = toupper(keyword[order[i]]);
            char cj = toupper(keyword[order[j]]);
            
            if (ci > cj || (ci == cj && order[i] > order[j])) {
                int temp = order[i];
                order[i] = order[j];
                order[j] = temp;
            }
        }
    }
}

void row_transposition_encrypt(const char *plaintext, const char *keyword, char *result) {
    int key_len = strlen(keyword);
    
    int padded_len = ((strlen(plaintext) + key_len - 1) / key_len) * key_len;
    char buf[100];
    
    for (int i = 0; i < strlen(plaintext); i++) {
        buf[i] = plaintext[i];
    }
    for (int i = strlen(plaintext); i < padded_len; i++) {
        buf[i] = 'X';
    }
    buf[padded_len] = '\0';
    
    int num_rows = padded_len / key_len;
    char grid[10][20];
    
    for (int i = 0; i < num_rows; i++) {
        for (int j = 0; j < key_len; j++) {
            grid[i][j] = buf[i * key_len + j];
        }
    }
    
    int order[20];
    get_key_order(keyword, order);
    
    int r_idx = 0;
    for (int col = 0; col < key_len; col++) {
        int orig_col = -1;
        for (int i = 0; i < key_len; i++) {
            if (order[i] == col) {
                orig_col = i;
                break;
            }
        }
        
        for (int row = 0; row < num_rows; row++) {
            result[r_idx] = grid[row][orig_col];
            r_idx++;
        }
    }
    result[r_idx] = '\0';
}

void row_transposition_decrypt(const char *ciphertext, const char *keyword, char *result) {
    int key_len = strlen(keyword);
    int cipher_len = strlen(ciphertext);
    int num_rows = (cipher_len + key_len - 1) / key_len;
    
    int order[20];
    get_key_order(keyword, order);
    
    char grid[10][20];
    for (int i = 0; i < num_rows; i++) {
        for (int j = 0; j < key_len; j++) {
            grid[i][j] = '\0';
        }
    }
    
    int char_idx = 0;
    for (int new_col = 0; new_col < key_len; new_col++) {
        int orig_col = order[new_col];
        for (int row = 0; row < num_rows; row++) {
            if (char_idx < cipher_len) {
                grid[row][orig_col] = ciphertext[char_idx];
                char_idx++;
            }
        }
    }
    
    int r_idx = 0;
    for (int i = 0; i < num_rows; i++) {
        for (int j = 0; j < key_len && grid[i][j] != '\0'; j++) {
            result[r_idx] = grid[i][j];
            r_idx++;
        }
    }
    
    while (r_idx > 0 && result[r_idx - 1] == 'X') {
        r_idx--;
    }
    result[r_idx] = '\0';
}

int main() {
    const char *text = "ATTACK AT DAWN";
    const char *key = "ZEBRAS";
    
    char encrypted[100];
    char decrypted[100];
    
    row_transposition_encrypt(text, key, encrypted);
    row_transposition_decrypt(encrypted, key, decrypted);
    
    printf("Original: %s\n", text);
    printf("Key: %s\n", key);
    printf("Encrypted: %s\n", encrypted);
    printf("Decrypted: %s\n", decrypted);
    
    return 0;
}
